### PR TITLE
Datastore revamp 5: `DataStore::to_data_tables()`

### LINKED ISSUE #1758: Make `RowId` & `TableId` independent types and remove `MsgId`
What the title says.

### PR DESCRIPTION
**Pull request by teh-cmc**
A prerequisite for the upcoming work on `LogDb`.

This is now extremely trivial to do with all the latest changes (the store is already a table, msgid mismatch is gone...).

---

1. #1727 
2. #1735
3. #1739 
4. #1785 
5. #1791 
6. #1795
7. #1801

---

On top of #1785 
Part of #1619 

Future work:
- #1793 
- #1794 

---

- [x] Self-review
- [x] No bench regression
- [x] No test regression
- [x] TODO all the things that fit into #1619 
- [x] open issues for everything else

### CODE REVIEW

File: crates/re_arrow_store/src/store_dump.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+use crate::{
+    store::{IndexedBucketInner, PersistentIndexedTable},
+    DataStore, IndexedBucket,
+};
+
+// ---
+
+impl DataStore {
+    /// Serializes the entire datastore into an iterator of [`DataTable`]s.
+    // TODO(cmc): This shouldn't dump cluster keys that were autogenerated.
```
Comment (teh-cmc) [inline crates/re_arrow_store/src/store_dump.rs:12]: issue

File: crates/re_arrow_store/src/store_stats.rs [hunk 2]
```diff
@@ -2,13 +2,12 @@ use nohash_hasher::IntMap;
...
 use crate::{
-    store::IndexedBucketInner, DataStore, DataStoreConfig, IndexedBucket, IndexedTable,
-    PersistentIndexedTable,
+    store::IndexedBucketInner, DataStore, IndexedBucket, IndexedTable, PersistentIndexedTable,
 };
 
 // ---
 
-#[derive(Default, Debug, Clone)]
+#[derive(Default, Debug, Clone, PartialEq, Eq, PartialOrd)]
```
Comment (emilk) [inline crates/re_arrow_store/src/store_stats.rs:10]: does it make sense to order the stats?

File: crates/re_arrow_store/tests/dump.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+        let mut store1 = DataStore::new(InstanceKey::name(), config.clone());
+        let mut store2 = DataStore::new(InstanceKey::name(), config.clone());
+        let mut store3 = DataStore::new(InstanceKey::name(), config.clone());
+
+        data_store_dump_impl(&mut store1, &mut store2, &mut store3);
+    }
+}
+
+fn data_store_dump_impl(store1: &mut DataStore, store2: &mut DataStore, store3: &mut DataStore) {
```
Comment (emilk) [inline crates/re_arrow_store/tests/dump.rs:33]: Why pass in the three stores by `&mut ` when they must start out empty, and are not used outside the function?

```suggestion
        data_store_dump_impl(config);
    }
}

fn data_store_dump_impl(config) {
    let mut store1 = DataStore::new(InstanceKey::name(), config.clone());
    let mut store2 = DataStore::new(InstanceKey::name(), config.clone());
    let mut store3 = DataStore::new(InstanceKey::name(), config.clone());
}
```

**Review by emilk — APPROVED** [review]

File: crates/re_arrow_store/src/store_dump.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            .flatten()
+    }
+}
+
+fn generate_table_id() -> TableId {
+    // NOTE: The `TableId` is only used for debugging purposes so this doesn't matter
+    // much... also we don't support saving to file on the web anyhow.
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        TableId::random()
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_arrow_store/src/store_dump.rs:193]: PR: https://github.com/rerun-io/rerun/pull/1796

File: crates/re_arrow_store/src/store_stats.rs (hunk 2 above)
Comment (teh-cmc) [reply crates/re_arrow_store/src/store_stats.rs:10]: used in tests to check some basic invariants

File: crates/re_arrow_store/tests/dump.rs (hunk 3 above)
Comment (teh-cmc) [reply crates/re_arrow_store/tests/dump.rs:33]: 1. Consistency with other test suites
2. I like keeping initialization and config permutation logic separate from the data test

**Pull request merged**